src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.String.cs

```diff
@@ -52,31 +52,60 @@ public static TValue Deserialize<TValue>(string json, JsonSerializerOptions opti
         public static object Deserialize(string json, Type returnType, JsonSerializerOptions options = null)
         {
             if (json == null)
+            {
                 throw new ArgumentNullException(nameof(json));
+            }
 
             if (returnType == null)
+            {
                 throw new ArgumentNullException(nameof(returnType));
+            }
 
-            return ParseCore(json, returnType, options);
-        }
-
-        private static object ParseCore(string json, Type returnType, JsonSerializerOptions options = null)
-        {
             if (options == null)
             {
                 options = JsonSerializerOptions.s_defaultOptions;
             }
 
-            // todo: use an array pool here for smaller requests to avoid the alloc?
-            byte[] jsonBytes = JsonReaderHelper.s_utf8Encoding.GetBytes(json);
-            var readerState = new JsonReaderState(options.GetReaderOptions());
-            var reader = new Utf8JsonReader(jsonBytes, isFinalBlock: true, readerState);
-            object result = ReadCore(returnType, options, ref reader);
+            object result;
+            byte[] tempArray = null;
 
-            if (reader.BytesConsumed != jsonBytes.Length)
+            int maxByteCount;
+            if (json.Length > JsonConstants.MaxEscapedTokenSize / JsonConstants.MaxExpansionFactorWhileTranscoding)
+            {
+                // Get the actual byte count in order to handle large input.
+                maxByteCount = JsonReaderHelper.GetUtf8ByteCount(json);
+            }
+            else
+            {
+                maxByteCount = json.Length * JsonConstants.MaxExpansionFactorWhileTranscoding;
+            }
+
+            Span<byte> utf8 = maxByteCount <= JsonConstants.StackallocThreshold ?
+                stackalloc byte[maxByteCount] :
+                (tempArray = ArrayPool<byte>.Shared.Rent(maxByteCount));
+
+            try
+            {
+                int actualByteCount = JsonReaderHelper.GetUtf8FromText(json, utf8);
```

Comment:
If we are going to use these helpers outside of the `Utf8JsonReader`, let's move it to the `JsonHelpers` class (which contains helpers used across multiple components) - this clean up can be done separately to avoid unnecessary code churn for a PR that is meant for 3.1.